src/referenced-fields.ts

```diff
@@ -294,3 +322,302 @@ export const getModifiedFields: EngineInstance['getModifiedFields'] = (
+			}
+
+			return findTableInTableDefinition(joinNode[1], candidates);
+
+		case 'Where':
+			const whereNode = statement as WhereNode;
+			findTablesInConstraint(whereNode[1], candidates);
+
+			return null;
+
+		default:
+			return null;
+	}
+};
+
+const findBindCandidates = (
+	statement: AbstractSqlQuery,
+	candidates: BindCandidate[],
+) => {
+	if (!['SelectQuery', 'UnionQuery'].includes(statement[0])) {
+		return;
+	}
+
+	switch (statement[0]) {
+		case 'SelectQuery':
+			const selectQueryNode = statement as SelectQueryNode;
+			const scope = statement;
+			for (const arg of selectQueryNode.slice(1)) {
+				assertAbstractSqlIsNotLegacy(arg);
+				const table = findTableInStatement(arg, candidates);
+				if (table) {
+					candidates.push({ ...table, scope });
+				}
+			}
+
+			break;
+
+		default:
+			throw new Error(
+				`findBindCandidates only supports SELECT queries, got ${statement[0]}`,
+			);
+	}
+};
+
+// TODO:
+// - This function only narrows the root table of the rule. This is always
+//   safe when the root table isn't selected from more than once and it is not
+//   negated in the LF. Right now we conservatively check for the former but
+//   not the second. The negative forms (e.g. it is forbidden that ...) are
+//   not fully supported anyway.
+// - Removing multiple candidates selecting from the same database table to
+//   avoid visibility issues is too conservative. The correct criteria is to
+//   just remove any that are present in at least 2 disjoint subqueries.
+//   Because in this case the problem is that in those cases there is not a
+//   single place in the query that has visibility inside both disjoint
+//   subqueries and that is a requirement for adding the corrent binds for
+//   narrowing.
+// - We can miss subqueries as we don't recurse into everything.
+// - Ban on aggregates is not necessary but solving it requires additional
+//   complexity.
+// - We only care about and recognize `COUNT` aggregates and `GROUP BY`
+//   statements right now.
+// - We assume the ID column is named "id".
+// - This is a very restricted implementation of narrowing which could be
+//   expanded to cover more situations.
+// - Only `SELECT` queries are supported while, theoretically, `UNION` queries
+//   are also valid.
+//
+// This function modifies `abstractSql` in place.
+export const addAffectedIdsBinds = (
+	abstractSql: AbstractSqlQuery,
+	lfRuleInfo: LfRuleInfo,
+) => {
+	const candidates: BindCandidate[] = [];
+	findBindCandidates(['SelectQuery', ['Where', abstractSql]], candidates);
+
+	// Count the number of references for each database table. Multiple
+	// candidates referencing the same table will be removed to conservatively
+	// avoid issues
+	const seenTableNames: { [key: string]: number } = {};
+	for (const candidate of candidates) {
+		if (candidate.tableName in seenTableNames) {
+			seenTableNames[candidate.tableName] += 1;
+		} else {
+			seenTableNames[candidate.tableName] = 1;
+		}
```

Comment:
I think this is slightly simpler using the `??=` operator to initialize to 0
```suggestion
		seenTableNames[candidate.tableName] ??= 0;
		seenTableNames[candidate.tableName] += 1;
```